nixos/modules/services/misc/rsync.nix

```diff
@@ -0,0 +1,157 @@
+    '';
+
+    package = lib.mkPackageOption pkgs "rsync" { };
+
+    commonArgs = lib.mkOption {
+      type = types.listOf types.str;
+      default = [ ];
+      example = [
+        "--verbose"
+        "--archive"
+        "--update"
+        "--delete"
+        "--mkpath"
+      ];
+      description = ''
+        Common arguments to pass to the rsync command.
+      '';
+    };
+
+    jobs = lib.mkOption {
+      description = ''
+        Synchronization jobs to run.
+      '';
+      default = { };
+      type = types.attrsOf (
+        types.submodule {
+          options = {
+            sources = lib.mkOption {
+              type = types.listOf types.str;
+              description = ''
+                Source directories.
+              '';
+              example = [
+                "/srv/src1/"
+                "/srv/src2/"
+              ];
+            };
+
+            destination = lib.mkOption {
+              type = types.str;
+              description = ''
+                Destination directory.
+              '';
+              example = "/srv/dst/";
+            };
+
+            extraArgs = lib.mkOption {
+              type = types.listOf types.str;
+              default = [ ];
+              example = [
+                "--verbose"
+              ];
+              description = ''
+                Extra arguments to pass to the rsync command.
+              '';
+            };
+
+            timerConfig = lib.mkOption {
+              type = lib.types.nullOr (lib.types.attrsOf unitOption);
+              default = {
+                OnCalendar = "daily";
+                Persistent = true;
+              };
+              description = ''
+                When to run the job.
+              '';
+            };
+
+            inhibitsSleep = lib.mkOption {
+              default = false;
+              type = lib.types.bool;
+              example = true;
+              description = ''
+                Prevent the system from sleeping while running the job.
+              '';
+            };
+          };
+        }
+      );
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    assertions = [
+      {
+        assertion = lib.all (job: job.sources != [ ]) (lib.attrValues cfg.jobs);
+        message = ''
+          At least one source directory must be provided to rsync.
+        '';
+      }
+    ];
+
+    systemd = lib.mkMerge (
+      lib.mapAttrsToList (
+        jobName: job:
+        let
+          systemdName = "rsync-job-${jobName}";
+          description = "rsync job ${jobName}";
+        in
+        {
+          timers.${systemdName} = {
+            wantedBy = [ "timers.target" ];
+            inherit description;
+            inherit (job) timerConfig;
+          };
+
+          services.${systemdName} = {
+            inherit description;
+
+            serviceConfig = {
+              Type = "oneshot";
+              User = "root";
+              Group = "root";
```

Comment:
I'm curious about this. Should it be `root` here?

Reply:
I'm not sure, but the best approach is probably to make this configurable, with `root` as the default.

Reply:
Yes that would be great!